# Repaso completo — Ejercicios de práctica

Escenario: eres analista de datos en una consultora. Te llegan tareas de dos clientes distintos: uno tiene una cadena de retail (Superstore) y otro gestiona una plataforma de alojamientos (Airbnb). Tu manager te manda los encargos por email. Cada sección es un encargo distinto.

Datasets disponibles:
- `data/external/train.csv` — ventas Superstore
- `data/external/Listings.csv` — listings Airbnb

No hay código de partida. Solo los enunciados.

---
## Encargo 1 — Limpieza básica

> *"Acabo de recibir un CSV con datos de ventas. Antes de usarlo para nada necesito saber en qué estado está. ¿Puedes revisarlo y dejarlo limpio?"*

**1.1** Carga `train.csv` e imprime su shape, sus tipos de dato y cuántos nulos hay por columna.

**1.2** `Order Date` y `Ship Date` llegaron como texto. Conviértelas a datetime. Verifica que el cambio fue correcto imprimiendo el dtype de ambas columnas.

**1.3** Comprueba si hay filas duplicadas. Si las hay, elimínalas e imprime el shape antes y después.

**1.4** Imprime los valores únicos de la columna `Region`. ¿Hay algún valor que parezca un error o inconsistencia de formato?

In [ ]:
# siempre al importar pandas es una buena práctica no solo para el analista
# sino para los demás compañeros de grupo, la idea es estandarizar los métodos
# y seguir una línea común para todo el equipo. Primero importo la librería de pandas
import pandas as pd
# Para no importat el modulo entero de la librería, se importa principalmente una clase 
# que vive dentro de ese módulo, no generar ruído, y resumir el código es fundamental siempre 
# que el código funcione perfectamente
from pathlib import Path

def find_project_root():
    for parent in [Path.cwd()]+list(Path.cwd().parents):
        if (parent / 'data').exists():
            return parent
ROOT = find_project_root()
df = pd.read_csv(ROOT / 'data' / 'external' / 'train.csv', low_memory=False)
print("------------------------------------")
print("Limpieza básica y Extracción")
print("------------------------------------")
#Shape
print("Shape conformado por:")
print(df.shape)
print("rows y atributos")
print("------------------------------------")
# limpieza previa antes de limpioar cualquier cosa, transformar objetos en sus dtypes correctamente
# crear una copia
df = df.copy()

df['Order Date'] = pd.to_datetime(df['Order Date'], format='%d/%m/%Y')
df['Ship Date'] = pd.to_datetime(df['Ship Date'], format='%d/%m/%Y')


print('Tipos corregidos:')
print(df[['Order Date','Ship Date']].dtypes)


print("------------------------------------")
n_dup = df.duplicated().sum()
print(f"Filas duplicadas: {n_dup}")

if n_dup > 0:
    df = df.drop_duplicates
    print(f'Shape tras borrado de duplicados: {df.shape}')
else:
    print('Sin duplicados - no se modifica el DataFrame')

print("------------------------------------")
print("Inspección contreta en columna *Region*: ")
print(df['Region'].unique())
print("------------------------------------")
print("Demonstración del dataset: ")
df.head(10)


------------------------------------
Limpieza básica y Extracción
------------------------------------
Shape conformado por:
(9800, 18)
rows y atributos
------------------------------------
Tipos corregidos:
Order Date    datetime64[us]
Ship Date     datetime64[us]
dtype: object
------------------------------------
Filas duplicadas: 0
Sin duplicados - no se modifica el DataFrame
------------------------------------
Inspección contreta en columna *Region*: 
<StringArray>
['South', 'West', 'Central', 'East']
Length: 4, dtype: str
------------------------------------
Demonstración del dataset: 


,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,State,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales
0,1,CA-2017-152156,2017-11-08,2017-11-11,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,Kentucky,42420.0,South,FUR-BO-10001798,Furniture,Bookcases,Bush Somerset Collection Bookcase,261.9600
1,2,CA-2017-152156,2017-11-08,2017-11-11,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,Kentucky,42420.0,South,FUR-CH-10000454,Furniture,Chairs,"Hon Deluxe Fabric Upholstered Stacking Chairs,...",731.9400
2,3,CA-2017-138688,2017-06-12,2017-06-16,Second Class,DV-13045,Darrin Van Huff,Corporate,United States,Los Angeles,California,90036.0,West,OFF-LA-10000240,Office Supplies,Labels,Self-Adhesive Address Labels for Typewriters b...,14.6200
3,4,US-2016-108966,2016-10-11,2016-10-18,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,Florida,33311.0,South,FUR-TA-10000577,Furniture,Tables,Bretford CR4500 Series Slim Rectangular Table,957.5775
4,5,US-2016-108966,2016-10-11,2016-10-18,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,Florida,33311.0,South,OFF-ST-10000760,Office Supplies,Storage,Eldon Fold 'N Roll Cart System,22.3680
5,6,CA-2015-115812,2015-06-09,2015-06-14,Standard Class,BH-11710,Brosina Hoffman,Consumer,United States,Los Angeles,California,90032.0,West,FUR-FU-10001487,Furniture,Furnishings,Eldon Expressions Wood and Plastic Desk Access...,48.8600
6,7,CA-2015-115812,2015-06-09,2015-06-14,Standard Class,BH-11710,Brosina Hoffman,Consumer,United States,Los Angeles,California,90032.0,West,OFF-AR-10002833,Office Supplies,Art,Newell 322,7.2800
7,8,CA-2015-115812,2015-06-09,2015-06-14,Standard Class,BH-11710,Brosina Hoffman,Consumer,United States,Los Angeles,California,90032.0,West,TEC-PH-10002275,Technology,Phones,Mitel 5320 IP Phone VoIP phone,907.1520
8,9,CA-2015-115812,2015-06-09,2015-06-14,Standard Class,BH-11710,Brosina Hoffman,Consumer,United States,Los Angeles,California,90032.0,West,OFF-BI-10003910,Office Supplies,Binders,DXL Angle-View Binders with Locking Rings by S...,18.5040
9,10,CA-2015-115812,2015-06-09,2015-06-14,Standard Class,BH-11710,Brosina Hoffman,Consumer,United States,Los Angeles,California,90032.0,West,OFF-AP-10002892,Office Supplies,Appliances,Belkin F5C206VTEL 6 Outlet Surge,114.9000


---
## Encargo 2 — Limpieza profesional Airbnb

> *"El equipo de Airbnb nos pasó su export de listings. Necesito que lo limpies siguiendo un criterio profesional, no solo quitando nulos a lo bruto. Dame un resumen de qué decidiste en cada paso y por qué."*

**2.1** Carga `Listings.csv` con los parámetros correctos para que no falle la codificación. Imprime shape y tipos de dato.

**2.2** Aplica el orden correcto de limpieza. Para cada paso, justifica la decisión:
- Corrige los tipos de `host_since` y las columnas booleanas (`host_is_superhost`, etc.)
- Elimina o conserva duplicados
- Trata los nulos de `district` (86% nulos), `bedrooms` (10% nulos) y `host_response_rate` (46% nulos)
- Decide qué hacer con los nulos en `review_scores_rating` y justifica por qué no los rellenas

**2.3** Normaliza `host_response_time`: elimina espacios y convierte a minúsculas. Imprime los valores únicos antes y después.

**2.4** Crea la columna `amenities_count` a partir de la columna `amenities`. Imprime la media y el máximo de esa nueva columna.

In [ ]:
import pandas as pd

from pathlib import Path

def find_project_root():
    for parent in [Path.cwd()]+list(Path.cwd().parents):
        if (parent / 'data').exists():
            return parent
ROOT = find_project_root()
df = pd.read_csv(ROOT / 'data' / 'external' / 'Listings.csv', encoding='latin-1', low_memory=False)

df = df.copy()


print(f'Su shape: {df.shape}')
# Corrección de tipos
df['host_since'] = pd.to_datetime(df['host_since'])#,format='%d/%m/%Y'
print(f'Corrección de formatos: {df['host_since'].dtypes}')

col_bool = ['instant_bookable','host_is_superhost','host_has_profile_pic','host_identity_verified']
for col in col_bool:
    df[col] = df[col].map({'t': True, 'f': False})
print(df[col_bool].dtypes)

n_dup = df.duplicated().sum()
print(f'Filas duplicadas: {n_dup}')

if n_dup>0:
    df = df.drop_duplicates()
    print(f'Shape tras eliminar Duplicados: {df.shape}')
else:
    print('No existen duplicados')

#------------------------------------- Buen metodo para identificar porcentaje de nulos
nulos = (df.isnull().mean()*100).round(3)
print('Porcentaje de nulos por columna:')
print(nulos[nulos>0].to_string())
#-------------------------------------
UMBRAL_NULOS=0.50
col_drop_null = [
    c for c in df.columns 
    if df[c].isnull().mean()>UMBRAL_NULOS
    ]
print(f'Columnas con >{UMBRAL_NULOS*100:.0f}% nulos: {col_drop_null}')
df = df.drop(columns=col_drop_null)
print(f'Shape tras eliminar columnas vacías: {df.shape}')




In [ ]:
#pd.set_option('display.max_columns', None)
#df.head()

#pd.reset_option('display.max_columns')

---
## Encargo 3 — Outliers y correlación

> *"Antes de modelar precios de Airbnb necesito entender qué tan ruidosos son los datos. ¿Hay precios que no tienen sentido? ¿Qué variables se relacionan entre sí?"*

**3.1** Carga el dataset de Airbnb limpio. Imprime la distribución de `price`: mínimo, máximo, media, mediana y skewness. ¿La distribución es simétrica o asimétrica?

**3.2** Detecta outliers en `price` usando el método IQR. Imprime Q1, Q3, IQR, límite inferior, límite superior y cuántas filas quedan fuera de esos límites.

**3.3** Detecta outliers en `price` usando z-score (umbral 3). Compara el número de outliers detectados con el método IQR. ¿Cuál detecta más? ¿Por qué?

**3.4** Selecciona al menos 5 columnas numéricas relevantes y calcula la matriz de correlación. Imprime los pares con correlación mayor a 0.5 en valor absoluto.

**3.5** Haz un boxplot de `price` con los límites IQR marcados como líneas de referencia (puedes usar plotly). ¿Qué observas?

---
## Encargo 4 — Agrupaciones y agregaciones

> *"El cliente de Superstore quiere un resumen de ventas por distintos cortes. Necesito tablas que pueda llevar a una reunión esta tarde."*

**4.1** Calcula para cada `Region`: revenue total, ticket medio (media de Sales), número de órdenes únicas y número de clientes únicos. Ordena de mayor a menor revenue.

**4.2** Para cada combinación de `Region` y `Category`, calcula el revenue total. Presenta el resultado como una tabla cruzada (pivot) con las regiones en filas y las categorías en columnas.

**4.3** Añade al DataFrame original una columna `media_categoria` con la media de Sales de la categoría a la que pertenece cada fila (sin reducir el número de filas). Luego crea `desviacion_vs_categoria` = Sales menos esa media.

**4.4** ¿Cuáles son los 5 clientes con mayor revenue total? Para cada uno, muestra también cuántas órdenes hicieron y cuál fue su ticket medio.

---
## Encargo 5 — EDA dirigido con series temporales

> *"Necesito entender cómo han evolucionado las ventas en el tiempo. ¿Hay estacionalidad? ¿Qué trimestre fue el mejor históricamente?"*

**5.1** Agrupa las ventas por mes (conservando el año) y grafica la evolución mensual de revenue. ¿Ves algún patrón estacional?

**5.2** Calcula el revenue por trimestre y año. ¿Qué trimestre es consistentemente el más fuerte?

**5.3** Añade una columna de media móvil de 3 meses al revenue mensual. Grafícala junto a la serie original. ¿Qué aporta la media móvil que la serie original no muestra?

**5.4** Calcula la variación porcentual de revenue mes a mes (`pct_change`). ¿En qué mes hubo el mayor caída? ¿Y el mayor crecimiento?

**5.5** Calcula los días de envío (`Ship Date - Order Date`). ¿Cuál es la media por `Ship Mode`? ¿Hay algún modo de envío con tiempos sospechosos?

---
## Encargo 6 — Patrones y narrativa analítica

> *"Ya tenemos los números. Ahora necesito que me cuentes qué está pasando en el negocio, no que me pases tablas. Identifica los patrones más relevantes y explícalos."*

**6.1** Identifica la sub-categoría con mayor crecimiento de revenue entre el primer y el último año del dataset. ¿Cuál cayó más?

**6.2** Compara el ticket medio entre los tres segmentos de clientes (`Segment`). ¿Qué segmento gasta más por orden? ¿Ese mismo segmento también compra más veces?

**6.3** ¿Hay alguna región donde las ventas de una categoría específica sean desproporcionadamente altas comparado con las demás regiones? Identifica el caso más llamativo.

**6.4** Construye una observación analítica con esta estructura para al menos uno de los hallazgos anteriores:
- **Observación**: qué ves en los datos
- **Patrón**: qué tendencia o diferencia explica ese dato
- **Interpretación**: por qué crees que ocurre
- **Implicación**: qué debería hacer el negocio con esa información

---
## Encargo 7 — Repaso con Online Retail

> *"Tenemos un cliente nuevo del sector retail online. Adjunto un dataset de transacciones. Necesito respuestas rápidas antes del lunes."*

Para este encargo usa el dataset **Online Retail II** si lo tienes disponible, o en su defecto vuelve a usar `train.csv` adaptando las preguntas.

**7.1** ¿Cuántas transacciones hay? ¿Cuántas son cancelaciones (InvoiceNo empieza por 'C')? ¿Qué porcentaje del total representan?

**7.2** Elimina las cancelaciones y las filas con `CustomerID` nulo. ¿Cuántas filas quedan?

**7.3** Calcula el revenue de cada transacción (`Quantity × UnitPrice`). ¿Cuál es el país con mayor revenue total?

**7.4** ¿Cuáles son los 10 productos más vendidos por cantidad? ¿Y los 10 con mayor revenue?

**7.5** ¿Hay productos con `UnitPrice` igual a 0? ¿Cuántos? ¿Qué harías con ellos?

---
## Encargo 8 — Segmentación RFM + Diagnóstico de negocio

> *"Quiero saber quiénes son nuestros mejores clientes y cuáles están a punto de irse. Además necesito un diagnóstico rápido del año pasado vs el anterior."*

### Parte A — RFM

**8.1** Calcula Recency, Frequency y Monetary para cada cliente en `train.csv`. Usa como fecha de referencia el día siguiente al último pedido del dataset.

**8.2** Puntúa cada dimensión del 1 al 3 usando `pd.qcut()`. Recuerda que en Recency el score se invierte: menos días = score más alto.

**8.3** Crea una columna `rfm_score` sumando R + F + M. ¿Cuántos clientes tienen score 9 (el máximo)? ¿Cuántos tienen score 3 (el mínimo)?

**8.4** Clasifica cada cliente en un segmento (puedes definir tus propias reglas). Presenta una tabla con: segmento, número de clientes, revenue total del segmento y recency media.

**8.5** ¿Qué segmento debería recibir atención urgente? Justifica con los datos.

### Parte B — Diagnóstico de negocio

**8.6** Calcula revenue, clientes únicos, órdenes y ticket medio por año.

**8.7** ¿El revenue creció o cayó entre el penúltimo y último año? ¿Cuánto en porcentaje?

**8.8** Descompón el cambio: ¿es un problema de adquisición (menos clientes) o de monetización (menor ticket medio)?

**8.9** ¿Qué categoría creció más? ¿Qué región tuvo peor evolución?

**8.10** Escribe el diagnóstico final con esta estructura:
> *"El revenue [subió/bajó] un X% entre [año] y [año] principalmente porque [causa]. La categoría más afectada fue [X] con un [Y]% de variación. Se recomienda [acción concreta]."*

---
## Notas de trabajo

- Si un ejercicio no sale a la primera, anota qué intentaste y dónde te bloqueaste
- No uses soluciones de internet antes de intentarlo al menos 10 minutos
- Cuando termines una sección completa, vuelve a leer tu código y pregúntate: ¿puedo hacerlo más corto? ¿más claro?

**Símbolos de ayuda cuando preguntes:**
- `*` si estás bloqueado y no sabes ni por dónde empezar
- `@` si sabes qué quieres pero no recuerdas la sintaxis
- `!` si el código corre pero el resultado no tiene sentido
- `%` si hay un concepto que no entiendes